spec/models/product_spec.rb

```diff
@@ -65,4 +65,54 @@
       it { is_expected.to eq('42') }
     end
   end
+
+  describe '#extensions' do
+    let(:base_a) { FactoryGirl.create(:product) }
+    let(:base_b) { FactoryGirl.create(:product) }
+
+    let(:level_one_extension) { FactoryGirl.create(:product, :extension, base_products: [base_a, base_b]) }
+
+    let!(:level_two_extension_a) { FactoryGirl.create(:product, :extension, base_products: [level_one_extension], root_product: base_a) }
+    let!(:level_two_extension_b) { FactoryGirl.create(:product, :extension, base_products: [level_one_extension], root_product: base_b) }
+
+    it 'returns correct extensions for base A' do
+      expect(level_one_extension.extensions.for_root_product(base_a)).to eq([level_two_extension_a])
+    end
+
+    it 'returns correct extensions for base B' do
+      expect(level_one_extension.extensions.for_root_product(base_b)).to eq([level_two_extension_b])
+    end
+  end
+
+  describe '#mirrored_extensions' do
+    let(:base_a) { FactoryGirl.create(:product) }
+    let(:base_b) { FactoryGirl.create(:product) }
+
+    let(:level_one_extension) { FactoryGirl.create(:product, :extension, base_products: [base_a, base_b]) }
+
+    let(:level_two_extension_a_not_mirrored) { FactoryGirl.create(:product, :extension, base_products: [level_one_extension], root_product: base_a) }
+    let(:level_two_extension_b_not_mirrored) { FactoryGirl.create(:product, :extension, base_products: [level_one_extension], root_product: base_b) }
+
+    let(:level_two_extension_a_mirrored) do
+      FactoryGirl.create(:product, :extension, :with_mirrored_repositories, base_products: [level_one_extension], root_product: base_a)
+    end
+    let(:level_two_extension_b_mirrored) do
+      FactoryGirl.create(:product, :extension, :with_mirrored_repositories, base_products: [level_one_extension], root_product: base_b)
+    end
+
+    before do
+      level_two_extension_a_not_mirrored
+      level_two_extension_b_not_mirrored
+      level_two_extension_a_mirrored
+      level_two_extension_b_mirrored
+    end
```

Comment:
This looks funny, why not use `let!`?

Reply:
Because Rubocop complains about it:

```
spec/models/product_spec.rb:93:5: C: RSpec/LetSetup: Do not use let! for test setup.
    let!(:level_two_extension_a_not_mirrored) { FactoryGirl.create(:product, :extension, base_products: [level_one_extension], root_product: base_a) }
```

I guess using the `before` block makes the initialization order more explicit.
But I can disable the check if you think it doesn't make sense.

Reply:
Hmm strange rule. It must be new. We use `let!` all over the place (even in RMT hehe(

Reply:
[It seems it doesn't complain](https://github.com/backus/rubocop-rspec/blob/master/lib/rubocop/cop/rspec/let_setup.rb#L42) about all of the `let!` calls, but only about the ones that aren't called from anywhere.

Reply:
[Here's](https://github.com/backus/rubocop-rspec/issues/94) the discussion about why it's useful.

Reply:
I see. With that logic, they actual `create` calls should be moved into the `before` block then.

Reply:
Moving into the `create` block would make it impossible to reference them later in the `expect` block.
A couple of those aren't referenced, so yes, those could be moved in, but I'd say it's not very important.

Reply:
Ok, let's leave it as is then.

Reply:
@thutterer ^^^